docker/iroha-builder/Dockerfile

```diff
@@ -0,0 +1,136 @@
+FROM ubuntu:20.04
+
+# number of concurrent threads during build
+# usage: docker build --build-arg PARALLELISM=8 -t name/name .
+ARG PARALLELISM=1
+
+ENV IROHA_HOME /opt/iroha
+ENV IROHA_BUILD /opt/iroha/build
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+    apt-get -y --no-install-recommends install \
+        apt-utils software-properties-common wget gpg-agent \
+        libtool
+
+RUN apt-get update && \
+    apt-get -y --no-install-recommends install \
+        # compilers (gcc-9, gcc-10)
+        build-essential g++-9 g++-10 cmake ninja-build \
+        gdb gdbserver \
+        # CI dependencies
+        git ssh tar gzip ca-certificates gnupg \
+        # code coverage
+        lcov
+
+RUN apt-get update && \
+    apt-get -y --no-install-recommends install \
+        # Python3
+        python3-dev python3-pip python-is-python3 \
+        # other
+        curl file ccache libssl-dev \
+        gcovr cppcheck doxygen rsync graphviz graphviz-dev vim zip unzip pkg-config \
+        jq
+
+RUN apt-get update && \
+    apt-get -y --no-install-recommends install \
+        postgresql postgresql-contrib
+
+# compiler clang-10 and libc++ only on x86_64, for debug purpose
+RUN if [ `uname -m` = "x86_64" ]; then \
+      apt-get update && \
+      apt-get -y --no-install-recommends install \
+        clang-10 lldb-10 lld-10 libc++-10-dev libc++abi-10-dev clang-format; \
+    fi
+
+## Finish APT installation, clean up
+RUN apt-get -y clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    rm -rf /var/cache/apt/archives/*
```

Comment:
Do you squash docker images afterwards? Because there is no need otherwise to cleanup apt cache as it will be still in docker image layers.

If no, you can merge this layers together.

Reply:
You can delete this step, this looks like a mistake.

Reply:
Best practices say this should be https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

Sure, you have a ratio to merge all apt-get steps in one. And definitely sure clean step should be a part of install step. This is to be fixed.

I intentionally separated packages into different steps to be able to reuse layers when one of them is changed. I.e. do not reinstall all the packages listed. But take previous layer with most packages installed and add yet another ones. This is a kind of optimization when change Dockerfile often. 

After all job is done, I believe I should merge all apt steps in one. But if someone will add just one package, the whole layer will be rebuilt. I will take much more time and traffic. And this is a point to discuss.
But from the other side, this is a rare procedure - rebuild docker image.

Reply:
I do not think we should squash layers. At least base Ubuntu should be a separate layer.

Reply:
All apt-get merged in one step.